Update Getting Start With Rails guide to account for Dev Containers

andrewn617 opened the pull request:
### Motivation / Background

In Rails 7.2, new apps will be generated with a Dev Container already set up. We also introduced the [rails-new](https://github.com/rails/rails-new) tool which allows an app to be generated a machine that does not have Rails, or even Ruby, installed.

This PR updates the Getting Started with Rails guide to account for rails-new and working with Dev Containers in VS Code.

### Detail

I updated the "Installing Rails" section of this guide to instead give instructions for installing Docker, VScode and rails-new. The "Creating the Blog Application" has also been updated to show usage of `rails-new` instead of the `rails new` command. And, I added a section about opening the app in a dev container within VSCode.

I moved the old section about installing rails and ruby locally to it's own guide, Installing Rails. That guide then links back to the getting started guide's "Hello, Rails!" section so someone reading that guide can continue with the getting started guide once they have Rails installed.

The other option was to keep both sets of instructions (rails local install vs. dev container), but I found it a bit awkward since the "Creating the Blog Application" is slightly different, and the instructions for opening the container need to come after that. So it was hard to structure it without introducing a bunch of duplication, or adding links between sections of the guide. So the separate guide for installing Rails seemed better, but I am open to feedback.

### Additional information

<!-- Provide additional information such as benchmarks, references to other repositories, or alternative solutions. -->

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [ ] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

AmandaPerino commented:
@Ridhwana will take a look at this on Wednesday (not possible to add you as a reviewer, so commenting instead).

Ridhwana commented on guides/source/getting_started.md:
```diff
@@ -72,92 +72,63 @@ By following along with this guide, you'll create a Rails project called
-Rails requires Ruby version 3.1.0 or later. It is preferred to use the latest Ruby version.
-If the version number returned is less than that number (such as 2.3.7, or 1.8.7),
-you'll need to install a fresh copy of Ruby.
+### Installing rails-new
```
We have a sub-heading in this section also named `#### Installing rails-new` as well which feels a bit confusing. Maybe we could rename this main section to 

```suggestion
### Setup and Installation
```

What do you think?

Ridhwana commented on guides/source/getting_started.md:
```diff
@@ -72,92 +72,63 @@ By following along with this guide, you'll create a Rails project called
+If you would like to install Rails directly on your machine, you can see the
+[Installing Rails guide](installing-rails.html)
+
 NOTE: The examples below use `$` to represent your terminal prompt in a UNIX-like OS,
 though it may have been customized to appear differently. If you are using Windows,
 your prompt will look something like `C:\source_code>`.
```
With the re-arrangement of the sections, this note feels a little out of place before the installation section, maybe we could move it to after the intro paragraph in "Creating a Blog Application"?

Ridhwana commented on guides/source/getting_started.md:
```diff
@@ -72,92 +72,63 @@ By following along with this guide, you'll create a Rails project called
-Many popular UNIX-like OSes ship with an acceptable version of SQLite3.
-Others can find installation instructions at the [SQLite3 website](https://www.sqlite.org).
+Once Docker has been installed, open it from your application folder to begin running
+the Docker engine on your machine.
```
Maybe we can also add something like this to connect Docker and DevContainer: 

Dev Containers will leverage Docker's containers to allow you to encapsulate your development environment and dependencies within isolated containers.

Ridhwana commented on guides/source/getting_started.md:
```diff
@@ -72,92 +72,63 @@ By following along with this guide, you'll create a Rails project called
 
-To verify that you have everything installed correctly, you should be able to
-run the following in a new terminal:
+The easiest way of working with Dev Containers is using the VSCode editor. You can
+install VSCode by downloading it from [the website](https://code.visualstudio.com/).
```
Do you think we could add a sentence that describes how VS code is beneficial when working with Dev Containers. Maybe something like: 

```suggestion
VSCode's integration with Dev Containers helps to streamline your development workflow. It helps to quickly spin up containers, access terminal commands, debug code, and utilize extensions. You can install VSCode by downloading it from [the website](https://code.visualstudio.com/).
```

Ridhwana replied:
In addition, I thought we need to install the [Visual Code extension CLI](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers) to start and attach the development container? Maybe that's not correct though?

If it turns out wew do need the extension then we coul summarise this for folks to explain its purpose "The Visual Studio Code Dev Containers extension allows you to open any folder inside (or mounted into) a container and take advantage of Visual Studio Code's full feature set. A [devcontainer.json file](https://code.visualstudio.com/docs/devcontainers/containers#_create-a-devcontainerjson-file) in your project tells VS Code how to access (or create) a development container with a well-defined tool and runtime stack. You can install it by..."

andrewn617 replied:
Argh! Yes thank you I missed that step.

Ridhwana commented on guides/source/getting_started.md:
```diff
@@ -72,92 +72,63 @@ By following along with this guide, you'll create a Rails project called
+#### Installing rails-new
 
-If it says something like "Rails 7.2.0", you are ready to continue.
+`rails-new` is a tool which can generate a new Rails application without needing
+Ruby or Rails installed on your machine. To install rails-new, follow the installation
+instructions [in the readme](https://github.com/rails/rails-new?tab=readme-ov-file#installation).
```
Same as above, connecting up the installation tools could be useful?

```suggestion
`rails-new`  generates a new Rails application for you without having to install Ruby on your machine. It uses Docker to generate the Rails application, thus allowing Docker to take care of installing the correct Ruby and Rails versions for you.

To install rails-new, follow the installation instructions [in the readme](https://github.com/rails/rails-new?tab=readme-ov-file#installation).
```

Ridhwana commented on guides/source/getting_started.md:
```diff
@@ -194,6 +165,38 @@ of the files and folders that Rails creates by default:
+
+### Opening the Blog Application in a Dev Container
+
+A Dev Container is a fully functional development environment running in a container
```
I suggested explaining what a Dev Container is at the start of the guide when we mention Dev Containers for the first time, so maybe if we decide to go that route, we can remove this introduction of what a Dev Container is?

Ridhwana commented on guides/source/getting_started.md:
````diff
@@ -194,6 +165,38 @@ of the files and folders that Rails creates by default:
+on your machine. Our new Rails application comes with a Dev Container already
+configured and ready to use.
+
+We will use VSCode to spin up and work with our Dev Container. First we need to open
+our app in VSCode. This can be done by opening VSCode and then opening our application's
+folder, or directly from the command line within our application's directory with:
+
+```bash
+$ code .
+```
````
I'm not sure that we need to explain _how_ to open a file in VSCode as it feels a bit too verbose, maybe we can just say to launch the VSCode application? I don't feel strongly about this if you feel that its better to go into the detail.

AmandaPerino replied:
I think this is a good tip for new folks who don't yet know command line prompts. I say we keep it.

andrewn617 replied:
I actually just realized the `code` command is not added to your path automatically when you install vscode. So I am going to remove this, other wise I will need to explain how to create it - which involves opening vscode 😅 so it's kind of a side quest at that point.

andrewn617 replied:
But I tried to simplify it.

Ridhwana commented on guides/source/getting_started.md:
````diff
@@ -194,6 +165,38 @@ of the files and folders that Rails creates by default:
+```
+
+When we open the application, VSCode will recognize that a Dev Container is available
+and will prompt us to re-open the application in a Dev Container. Click the prompt
````
maybe we can allude to what the prompt is, something like "Folder contains a Dev Container configuration file. Reopen folder to develop in a container ([learn more](https://aka.ms/vscode-remote/docker))."

Ridhwana replied:
Maybe we should also add a reminder to have the Docker application running?

andrewn617 replied:
I was thinking to avoid matching the language exactly as VScode could change it. But I will make it a bit more clear.

Ridhwana replied:
Sounds good!

Ridhwana commented on guides/source/getting_started.md:
```diff
@@ -194,6 +165,38 @@ of the files and folders that Rails creates by default:
+and will prompt us to re-open the application in a Dev Container. Click the prompt
+and the Dev Container will begin initializing.
+
+Once the Dev Container setup is complete, VSCode will now be working with a fully
+setup development environment, with Ruby, Rails and all gems and dependencies installed.
```
Should this be?

```suggestion
Once the Dev Container setup is complete, your application will now be running in a fully
setup development environment; with Ruby, Rails, and all gems and dependencies installed.
```

Ridhwana replied:
What do you think about this rephrasing?

```suggestion
Once the Dev Container setup is complete, your development environment is ready to use, with Ruby, Rails, and all your dependencies installed. 
```

Ridhwana commented on guides/source/getting_started.md:
```diff
@@ -194,6 +165,38 @@ of the files and folders that Rails creates by default:
+Once the Dev Container setup is complete, VSCode will now be working with a fully
+setup development environment, with Ruby, Rails and all gems and dependencies installed.
+
+You can open the terminal within VScode to verify:
```
```suggestion
You can open the terminal within VSCode to verify that Rails is installed:
```

Ridhwana commented on guides/source/getting_started.md:
````diff
@@ -194,6 +165,38 @@ of the files and folders that Rails creates by default:
+Rails 7.2.0
+```
+
+For the remainder of this guide, you will be working from the terminal in VSCode, which
+is your entry point for working your Application running inside the Dev Container.
````
```suggestion
For the remainder of this guide, you will be working within VSCode which is your entry point for your Application running inside the Dev Container.
```

Ridhwana replied:
What do you think about? 

```suggestion
For the remainder of this guide, you will be working within VSCode which serves as your entry point for your application that will run in the Dev Container.
```

Ridhwana reviewed (commented):
Hi @andrewn617 this looks really good and it's great to see installation instructions using Dev Container! I left some suggestions and feedback, let me know if anything is unclear.

Ridhwana commented on guides/source/getting_started.md:
```diff
@@ -72,92 +72,63 @@ By following along with this guide, you'll create a Rails project called
-If it says something like "Rails 7.2.0", you are ready to continue.
+`rails-new` is a tool which can generate a new Rails application without needing
+Ruby or Rails installed on your machine. To install rails-new, follow the installation
+instructions [in the readme](https://github.com/rails/rails-new?tab=readme-ov-file#installation).
```
@andrewn617 I was just trying to install this locally and I don't see any installation instructions in the readme for `rails-new`?

andrewn617 replied:
I was chatting with Rafael about this today and it's next on my todo list. I'll have a PR open for it in the next 24 hours.

andrewn617 commented:
@Ridhwana Thank you for all your feedback! This is really helpful. I took most of your feedback.

One thing I am a bit unsure about: in a few places I had something like "working with your application in a container" and you recommended changing it to "run your application in a container". I avoided using the word "running" here because it makes me think of running the server. But the dev container running and the application running (within the container) are different. Maybe I am overthinking it and for the audience of the guide this distinction isn't important 🤔 but I wanted to get your thoughts on it. I was trying to stick with saying things "working with your application within the container" or "developing within the container" but I admit the results are a bit awkward.

Ridhwana commented:
> One thing I am a bit unsure about: in a few places I had something like "working with your application in a container" and you recommended changing it to "run your application in a container". I avoided using the word "running" here because it makes me think of running the server. But the dev container running and the application running (within the container) are different. I wonder how best to succinctly capture "the container is now running, it's a fully featured development environment from which you can edit code, run tests, boot up the console, or run the application server".

Ah, thank you for this clarification! I think it's a valid concern especially when folks are new to all of these tools, I'll go back to where I suggested you change it and see if there's a different type of wording.

Ridhwana commented on guides/source/getting_started.md:
```diff
@@ -7,8 +7,8 @@ This guide covers getting up and running with Ruby on Rails.
-* How to install Rails, create a new Rails application, and connect your
-  application to a database.
+* How create a new Rails application
+* How to work with your Rails application in a Dev Container
```
What about?

```suggestion
* How to create, set up, and run a Rails application (in a Dev Container)
```

andrewn617 commented:
@Ridhwana Okay, I have pushed a new version which covers all your comments. Take another look when you have a chance :)

Ridhwana commented:
> @Ridhwana Okay, I have pushed a new version which covers all your comments. Take another look when you have a chance :)

Thanks so much! I'll take a look tomorrow. Also, will you let me know once the installation instructions for rails-new are ready and then I'll test out the setup as well.

zzak commented on guides/source/getting_started.md:
````diff
@@ -72,92 +71,72 @@ By following along with this guide, you'll create a Rails project called
-ruby 3.1.0
-```
+Alternatively, to install Rails directly on your machine, you can read through the
+[Installing Rails guide](installing-rails.html)
````
```suggestion
[Installing Rails guide](installing_rails.html)
```

andrewn617 committed:
Update Getting Start With Rails guide to account for Dev Containers

Create a new guide called Getting Started with Dev Containers and link to it from the original Getting Started guide.

andrewn617 commented:
@Ridhwana Heads up, based on feedback from the Rails Core team I have inverted my approach to updating the guides. I created a new guide called Getting Started with Dev Containers and linked to it from the original getting started guide. The content has stayed pretty much the exact same.

Also I have a PR open with instructions for rails-new https://github.com/rails/rails-new/pull/13

Ridhwana commented:
@andrewn617 I tried installing it based on your instructions but I may doing something wrong. 

1. I downloaded the tar file from https://github.com/rails/rails-new/releases/tag/v0.3.0, not sure if it's okay to use this pre-release version in the meantime? 
2. Then I unzipped the file on my Macbook by using `tar -zxvf rails-new-0.3.0.tar.gz`
3. I moved the file to `usr/local/bin` using `sudo mv rails-new-0.3.0 /usr/local/bin/`  which moves the executable into my path 
4. Then I tried running `rails-new-0.3.0`.

Maybe I'm missing something obvious?

Ridhwana commented:
> @Ridhwana Ah, sorry. You need to download the executable for your platform (eg rails-new-aarch64-apple-darwin.tar.gz for mac m1), not the source code. I'll update the rails new instructions to be more clear on that.

Hey Andrew! Thanks for your guidance on this. I anticipate we might need to exchange a few messages as I work through the instructions to ensure everyone can smoothly install it without any hiccups. Here's my experience thus far:

I grabbed the executable from [this link](https://github.com/rails/rails-new/releases/tag/v0.3.0) for version `v0.3.0`. After that, I followed your steps and relocated the executable to `usr/local/bin`, which is on the executable path in my system.

Moving forward, I navigated to a different directory and executed `rails-new blog`. However, I encountered an error stating "'rails-new' can’t be opened because Apple cannot check it for malicious software." After doing some digging, it seems like macOS's Gatekeeper security feature is causing this snag. It's flagging the rails-new-0.3.0 executable due to it being from an unidentified developer or lacking notarization. Do you reckon this will be addressed in the latest release version?

To sidestep this roadblock temporarily and continue testing, I opted to override Gatekeeper using `xattr -d com.apple.quarantine /usr/local/bin/rails-new` to execute the file.

Afterwards, I retried `rails-new blog`, and it successfully generated a folder containing all the necessary files 🎉

Following that, I ensured my Docker engine was up and running, and then opened the folder in VSCode. However, I hit an issue. Here's a snapshot of the error message regarding .rbenv ruby version:

<img width="523" alt="Screenshot 2024-05-02 at 11 05 00" src="https://github.com/rails/rails/assets/2786819/82d64b4e-2abf-4d68-9195-b21ae5bd4051">

Looking forward to your insights on this issue. Thanks again for your help!

andrewn617 commented:
Hmm did you get that error right when you opened VSCode? Or after you started the container? Opening is VSCode does not automatically start the container, it should give you a pop up that prompts you to open the container:

<img width="539" alt="Screenshot 2024-05-02 at 6 44 29 AM" src="https://github.com/rails/rails/assets/39735028/07d82baf-d913-4fae-99d2-4a62d52f9673">

Ridhwana commented:
> Hmm did you get that error right when you opened VSCode? Or after you started the container? Opening is VSCode does not automatically start the container, it should give you a pop up that prompts you to open the container:

Does this help?

https://www.loom.com/share/5a0fe7ea1796498483592977531096e4?sid=3a2bea0b-2baa-460c-99bd-4c564869bf92

andrewn617 commented:
Oh! I see the issue. Rails 7.2 is not released yet but these instructions are written for that release 😅 So you need to add the `--main` flag when doing `rails-new` to get the `.devcontainer` folder. Otherwise it will be your rails gem version (7.1.x). If you try again with `rails-new blog --main` you should see the `.devcontainer`! Sorry about that!

Ridhwana commented on guides/source/getting_started_with_devcontainer.md:
````diff
@@ -0,0 +1,123 @@
+Rails 7.2.0
+```
+
+You can now continue with the [Getting Started guide](getting_started.md#hello-rails) and
+begin building your Blog application. You will be working within VSCode, which serves as
+your entry point to your application's Dev Container, where you can run code, run tests and
+run your application.
````
[This section](https://github.com/rails/rails/blob/4e3d2f7ea0ad5d939f82840c7ccc8db1a665eb1f/guides/source/getting_started.md#starting-up-the-web-server) in the [Getting Started guide](getting_started.md#hello-rails) talks about starting up the web server on your development machine, I'm wondering do we need a little more verbose instructions here about what happens once you get you able to open your blog application within Dev Container? Things like how do you run your server and see your application?

andrewn617 replied:
Nope, everything is the exact same. If you `bin/rails s` within the container, it will boot the app and be accessible in your browser at `http://127.0.0.1:3000`.

Ridhwana reviewed (commented):
@andrewn617 I managed to create a new application based on your instructions, thanks for help! 

My final questions and thoughts on this review after testing are: 
- Are we planning on waiting to merge this only after we release Rails 7.2 so that the instructions hold true or are we adding a disclaimer for versions earlier than 7.2? 
- Are we also waiting for the latest release to be released https://github.com/rails/rails-new/releases/tag/latest before merging or do we have a different plan in place?
- What can we do about Gatekeeper that doesn't allow us to run the executable without overriding it? 
- I dropped a comment about possibly mentioning what happens after opening up the blog application in DEV Container.

andrewn617 commented:
@Ridhwana 

> Are we planning on waiting to merge this only after we release Rails 7.2 so that the instructions hold true or are we adding a disclaimer for versions earlier than 7.2?

It's my first time contributing the guides, so my understanding might not be exactly right. But I believe once this merges, it will be on the ["edge" guides](https://edgeguides.rubyonrails.org/), which are the guides for the latest main branch of Rails. Once Rails 7.2 those guides will be promoted to the [official guides](https://guides.rubyonrails.org/) (which you can see now are for v7.1.3.2).

> Are we also waiting for the latest release to be released https://github.com/rails/rails-new/releases/tag/latest before merging or do we have a different plan in place?

This will have an official release prior to 7.2

> What can we do about Gatekeeper that doesn't allow us to run the executable without overriding it?

I believe Rafael is working on that. I will chat with him today about it

I'll answer your other question in line =)

Ridhwana commented:
Sounds good, thanks @andrewn617! ✨

@AmandaPerino just a heads up that I've reviewed this,  and the instructions have been tested too. [These are some of the considerations](https://github.com/rails/rails/pull/51637#issuecomment-2092925501) that may want to be noted.

Ridhwana commented on guides/source/getting_started_with_devcontainer.md:
```diff
@@ -0,0 +1,123 @@
+**DO NOT READ THIS FILE ON GITHUB, GUIDES ARE PUBLISHED ON https://guides.rubyonrails.org.**
```
@andrewn617 it just occurred to me that we need to [wrap the file to 80 columns](https://edgeguides.rubyonrails.org/ruby_on_rails_guides_guidelines.html#column-wrapping) if you have not done that already. 

Please exclude line 1 from this wrapping. Thanks!
